spec.md

```diff
@@ -1090,6 +1084,252 @@ message ControllerServiceCapability {
+
+Snapshot can be used as the source to provision a new volume.
+CreateVolumeRequest is modified to take an OPTIONAL parameter of a source snapshot.
+Reverting snapshot, where data in the original volume is erased and replaced with data in the snapshot, is an advanced functionality not every storage system can support and therefore is currently out of scope.
+
+```protobuf
+message CreateSnapshotRequest {
+  // The ID of the source volume to be snapshotted.
+  // This field is REQUIRED.
+  string source_volume_id = 1;
+
+  // The suggested name for the snapshot. This field is REQUIRED for
+  // idempotency.
+  string name = 2;
+
+  // Secrets required by plugin to complete snapshot creation request.
+  // This field is OPTIONAL.
+  map<string, string> create_snapshot_secrets = 3;
+
+  // Plugin specific parameters passed in as opaque key-value pairs.
+  // This field is OPTIONAL. The Plugin is responsible for parsing and
+  // validating these parameters. COs will treat these as opaque.
+  // Use cases for opaque parameters:
+  // - Specify a policy to automatically clean up the snapshot.
+  // - Specify an expiration date for the snapshot.
+  // - Specify whether the snapshot is readonly or read/write.
+  // - Specify if the snapshot should be replicated to some place.
+  // - Specify primary or secondary for replication systems that
+  //   support snapshotting only on primary.
+  map<string, string> parameters = 4;
+}
+
+message CreateSnapshotResponse {
+  // Contains all attributes of the newly created snapshot that are
+  // relevant to the CO along with information required by the Plugin
+  // to uniquely identify the snapshot. This field is REQUIRED.
+  Snapshot snapshot = 1;
+}
+
+// The information about a provisioned snapshot.
+message Snapshot {
+  // SnapshotSize is the disk size of the snapshot. The purpose of this
+  // field is to give CO guidance on how much space is needed to create
+  // a volume from this snapshot. The size of the volume must not be
+  // less than the size of the source snapshot. This field is OPTIONAL.
+  // If SnapshotSize is not set, it indicates that this size is unknown.
+  SnapshotSize size = 1;
+
+  // Contains identity information for the created snapshot. This field
+  // is REQUIRED. The identity information will be used by the CO in
+  // subsequent calls to refer to the provisioned snapshot.
+  string id = 2;
+
+  // Identity information for the source volume. Note that create
+  // snapshot from snapshot is not supported here so the source has to
+  // be a volume. This field is REQUIRED.
+  string source_volume_id = 3;
+
+  // Timestamp when the point-in-time snapshot is taken on the storage
+  // system. The format of this field should be a UnixNano time encoded
```

Comment:
is `UnixNano` a standard? it might be a golang API but CSI isn't tied to a specific lang - can we find a more language neutral term (once that's also well-defined)?

Reply:
I'll change UnixNano to Unix nanoseconds which can be retrieved by other language as well.